File: src/frontend/src/api/context.ts
/**
 * Context API - Frontend wrappers for RAG context retrieval.
 *
 * Covers:
 * - Build context pack for a chapter (/context/build/{chapter_id})
 * - Index chapter content (/context/index/{chapter_id})
 * - Query RAG index (/context/query)
 * - List indexed chunks (/context/chunks/{chapter_id})
 * - Delete chapter index (/context/chunks/{chapter_id})
 * - RAG index statistics (/context/stats)
 */

import { api } from "./request"

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ContextBuildRequest {
  max_chars?: number
}

export interface ContextBuildResponse {
  chapter_id: number
  sections: Record<string, unknown>
  meta: Record<string, unknown>
  weights: Record<string, number>
}

export interface ContextQueryRequest {
  query: string
  strategy?: "auto" | "graph_hybrid" | "bm25_fallback" | "hybrid" | "vector" | "bm25"
  top_k?: number
  chunk_type?: string
  chapter_id?: number
  center_entities?: string[]
}

export interface ContextQueryResult {
  chunk_id: string
  chapter_id: number
  scene_index: number
  content: string
  score: number
  source: string
  chunk_type?: string
  parent_chunk_id?: string
  source_file?: string
}

export interface ContextQueryResponse {
  query: string
  strategy: string
  results: ContextQueryResult[]
  total: number
  degraded: boolean
  degraded_reason?: string
}

export interface ContextChunkResponse {
  chunk_id: string
  chapter_id: number
  scene_index: number
  content: string
  chunk_type: string
  parent_chunk_id?: string
  source_file?: string
  created_at?: string
}

export interface ContextChunksResponse {
  chapter_id: number
  chunks: ContextChunkResponse[]
  total: number
}

export interface ContextIndexRequest {
  content: string
  summary?: string
  max_chunk_size?: number
  overlap?: number
}

export interface ContextIndexResponse {
  chapter_id: number
  stored: number
  total_chunks: number
  degraded: boolean
  degraded_reason?: string
}

export interface ContextStatsResponse {
  vectors: number
  terms: number
  max_chapter: number
}

// ---------------------------------------------------------------------------
// API Methods
// ---------------------------------------------------------------------------

/** Build a context pack for the specified chapter. */
export const buildContext = (
  chapterId: number,
  request: ContextBuildRequest = {}
): Promise<ContextBuildResponse> =>
  api.post<ContextBuildResponse>(`/context/build/${chapterId}`, request)

/** Index chapter content into the RAG store. */
export const indexChapter = (
  chapterId: number,
  request: ContextIndexRequest
): Promise<ContextIndexResponse> =>
  api.post<ContextIndexResponse>(`/context/index/${chapterId}`, request)

/** Query the RAG index with the specified strategy. */
export const queryContext = (
  request: ContextQueryRequest
): Promise<ContextQueryResponse> =>
  api.post<ContextQueryResponse>("/context/query", request)

/** List indexed chunks for a chapter. */
export const getChunks = (chapterId: number): Promise<ContextChunksResponse> =>
  api.get<ContextChunksResponse>(`/context/chunks/${chapterId}`)

/** Delete all indexed chunks for a chapter. */
export const deleteChunks = (
  chapterId: number
): Promise<{ chapter_id: number; deleted: number }> =>
  api.delete<{ chapter_id: number; deleted: number }>(`/context/chunks/${chapterId}`)

/** Get RAG index statistics. */
export const getContextStats = (): Promise<ContextStatsResponse> =>
  api.get<ContextStatsResponse>("/context/stats")

// ---------------------------------------------------------------------------
// Grouped API export
// ---------------------------------------------------------------------------

export const contextApi = {
  buildContext,
  indexChapter,
  queryContext,
  getChunks,
  deleteChunks,
  getContextStats,
}

export default contextApi

File: src/backend/app/schemas/context.py
"""Context / RAG schemas (Phase 1 Track B.1).

Aligned with frontend `src/frontend/src/api/context.ts`:
- ContextChunkOut / ContextChunkList
- IndexRequest / IndexResponse
- ContextStatsOut
- WeightsResponse / WeightsUpdate
"""
from __future__ import annotations

from typing import Optional

from pydantic import Field

from app.schemas.base import BaseSchema


# ---------------------------------------------------------------------------
# Chunks
# ---------------------------------------------------------------------------

class ContextChunkOut(BaseSchema):
    """Indexed text chunk for RAG retrieval."""

    chunk_id: str
    chapter_id: int
    scene_index: int
    content: str
    chunk_type: str = "text"
    parent_chunk_id: Optional[str] = None
    source_file: Optional[str] = None


class ContextChunkList(BaseSchema):
    """List of chunks for a chapter."""

    chapter_id: int
    chunks: list[ContextChunkOut]
    total: int


# ---------------------------------------------------------------------------
# Indexing
# ---------------------------------------------------------------------------

class IndexRequest(BaseSchema):
    """Request body for POST /context/{chapter_id}/index."""

    content: str = Field(..., min_length=1, description="Chapter text to chunk + index")
    summary: Optional[str] = None
    max_chunk_size: int = Field(default=500, ge=50, le=5000)
    overlap: int = Field(default=50, ge=0, le=500)


class IndexResponse(BaseSchema):
    """Response for POST /context/{chapter_id}/index."""

    chapter_id: int
    stored: int
    total_chunks: int
    degraded: bool = False
    degraded_reason: Optional[str] = None


# ---------------------------------------------------------------------------
# Stats
# ---------------------------------------------------------------------------

class ContextStatsOut(BaseSchema):
    """Per-chapter RAG index statistics."""

    chapter_id: int
    vectors: int
    terms: int
    max_scene_index: int


# ---------------------------------------------------------------------------
# Weights
# ---------------------------------------------------------------------------

class WeightsResponse(BaseSchema):
    """GET /context/weights response."""

    weights: dict[str, float]


class WeightsUpdate(BaseSchema):
    """PUT /context/weights body."""

    weights: dict[str, float] = Field(..., description="Keyed retrieval weights")
